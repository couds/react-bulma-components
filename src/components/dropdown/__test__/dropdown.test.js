import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { JSDOM } from 'jsdom';
import Dropdown from '..';

describe('Dropdown component', () => {
  beforeEach(() => {
    // eslint-disable-next-line
    global.window = new JSDOM('<body><div id="app-root"></div></body>').window;
  });
  it('Should Exist', () => {
    expect(Dropdown).toMatchSnapshot();
  });
  it('Should have dropdown classname', () => {
    const component = renderer.create(
      <Dropdown value="value" onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should add listener do document on mount', () => {
    const app = global.window.document.querySelector('#app-root');
    global.window.document.addEventListener = jest.fn();
    const component = mount(
      <Dropdown value="value" onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
      {
        attachTo: app,
      },
    );
    expect(window.document.addEventListener).toHaveBeenCalled();
    component.unmount();
  });
  it('Should concat Bulma class with classes in props', () => {
    const component = renderer.create(
      <Dropdown value="value" className="other-class" onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should have custom inline styles', () => {
    const component = renderer.create(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should have divider', () => {
    const component = renderer.create(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
        <Dropdown.Divider />
        <Dropdown.Item value="other">Other</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should be right-aligned when using "right" prop', () => {
    const component = renderer.create(
      <Dropdown right>
        <Dropdown.Item value="value">Item</Dropdown.Item>
        <Dropdown.Item value="other">Other</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('should also be right-aligned when using "align" prop', () => {
    const component = renderer.create(
      <Dropdown align="right">
        <Dropdown.Item value="value">Item</Dropdown.Item>
        <Dropdown.Item value="other">Other</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should appear above the dropdown button', () => {
    const component = renderer.create(
      <Dropdown up>
        <Dropdown.Item value="value">Item</Dropdown.Item>
        <Dropdown.Item value="other">Other</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should open the Dropdown', () => {
    const component = shallow(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
        <Dropdown.Divider />
        <Dropdown.Item value="other">Other</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.state('open')).toBe(false);
    component.find('.dropdown-trigger').simulate('click');
    expect(component.state('open')).toBe(true);
  });
  it('Should open the Dropdown and prevent default event (not to navigate if a link is on the dropdown trigger)', () => {
    const preventDefault = jest.fn();
    const component = shallow(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
        <Dropdown.Divider />
        <Dropdown.Item value="other">Other</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.state('open')).toBe(false);
    component.find('.dropdown-trigger').simulate('click', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
    expect(component.state('open')).toBe(true);
  });
  it('Should change the value', () => {
    const onChange = jest.fn();
    const component = shallow(
      <Dropdown value="" hoverable style={{ width: 400 }} onChange={onChange}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    component.find('.dropdown-trigger').simulate('click');
    component.find(Dropdown.Item).simulate('click');
    expect(onChange).toHaveBeenCalledWith('value');
    expect(component.state('open')).toBe(false);
  });
  it('Should close on select', () => {
    const component = mount(
      <Dropdown>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    component.find('.dropdown-trigger').simulate('click');
    component.find(Dropdown.Item).simulate('click');
    expect(component.state('open')).toBe(false);
  });
  it('Should close the dropdown', () => {
    const onChange = jest.fn();
    const component = mount(
      <Dropdown value="" style={{ width: 400 }} onChange={onChange}>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    component.find('.dropdown-trigger').simulate('click');
    component.find(Dropdown.Item).simulate('click', { path: [] });
    expect(component.state('open')).toBe(false);
  });
  it('Should show custom label passed to the label prop', () => {
    const component = renderer.create(
      <Dropdown label="test label">
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component).toMatchSnapshot();
  });
  it('Should show the label of the first dropdown item when no custom label is passed', () => {
    const component = renderer.create(
      <Dropdown>
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component).toMatchSnapshot();
  });
  it('Should show custom label when active valued is undefined/empty', () => {
    const component = shallow(
      <Dropdown label="test label" value="">
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.find('span').text()).toEqual('test label');
  });
  it('Should show the label of the dropdown item when value of it is the active value', () => {
    const component = shallow(
      <Dropdown label="test label" value="value">
        <Dropdown.Item value="value">Item</Dropdown.Item>
      </Dropdown>,
    );
    expect(component.find('span').text()).toEqual('Item');
  });
});
