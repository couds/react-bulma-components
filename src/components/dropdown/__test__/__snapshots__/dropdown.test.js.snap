// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Should Exist 1`] = `[Function]`;

exports[`Dropdown component Should concat Bulma class with classes in props 1`] = `
<div
  className="dropdown other-class"
  style={Object {}}
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <a
      className="button"
      disabled={false}
      href=""
      style={Object {}}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
        style={Object {}}
      >
        <i
          className="rbc rbc-angle-down"
        />
      </span>
    </a>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown component Should have custom inline styles 1`] = `
<div
  className="dropdown"
  style={
    Object {
      "width": 400,
    }
  }
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <a
      className="button"
      disabled={false}
      href=""
      style={Object {}}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
        style={Object {}}
      >
        <i
          className="rbc rbc-angle-down"
        />
      </span>
    </a>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown component Should have divider 1`] = `
<div
  className="dropdown"
  style={
    Object {
      "width": 400,
    }
  }
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <a
      className="button"
      disabled={false}
      href=""
      style={Object {}}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
        style={Object {}}
      >
        <i
          className="rbc rbc-angle-down"
        />
      </span>
    </a>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
      <hr
        className="dropdown-divider"
        style={Object {}}
      />
      <div
        className="dropdown-item"
        onClick={[Function]}
        role="presentation"
        title="other"
      >
        Other
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown component Should have dropdown classname 1`] = `
<div
  className="dropdown"
  style={Object {}}
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <a
      className="button"
      disabled={false}
      href=""
      style={Object {}}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
        style={Object {}}
      >
        <i
          className="rbc rbc-angle-down"
        />
      </span>
    </a>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
    </div>
  </div>
</div>
`;
